File: my_py_pkg/my_py_pkg/add_two_ints_client_no_oop.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts

def main(args=None):
    rclpy.init(args=args)
    node = Node("add_two_ints_client_no_oop")
    
    client = node.create_client(AddTwoInts, "add_two_ints") # make sure use the same service name
    # Wait until the service is up
    while not client.wait_for_service(1.0): # 1.0 second timeout
        node.get_logger().warn("Waiting for Add Two Ints server...")

    # Create a request object
    request = AddTwoInts.Request()
    request.a = 3
    request.b = 8

    # Send the request and get the response 
    future = client.call_async(request) # Non-blocking call, if we use client.call(request), it will block until get the response, because it's a synchronous call
                                        # future object is like the response from the server
    rclpy.spin_until_future_complete(node, future) # Spin until the future is complete
    response = future.result()
    node.get_logger().info(str(request.a) + " + " + str(request.b) + " = " + str(response.sum))


    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: my_cpp_pkg/src/add_two_ints_client_no_oop.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

using namespace std::chrono_literals;

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<rclcpp::Node>("add_two_ints_client_no_oop"); 

    auto client = node->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");

    // Wait for the service to be available
    while (!client->wait_for_service(1s)) 
    {
        RCLCPP_WARN(node->get_logger(), "Waiting for service to be available...");
    }
    // When the server is found, a shared pointer to a request is created, which is different from in the server cpp file, where we created an object
    auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
    
    request->a = 6;
    request->b = 2;

    auto future = client->async_send_request(request);
    rclcpp::spin_until_future_complete(node, future);

    auto response = future.get(); // The response is also a shared pointer
    RCLCPP_INFO(node->get_logger(), "%d + %d = %d",
                    (int)request->a, (int)request->b, (int)response->sum);

    rclcpp::shutdown();
    return 0;
}
